docs/design.md

```diff
@@ -0,0 +1,153 @@
+
+```yaml
+kind: Model
+spec:
+  container: # used for either loader or trainer
+  loader: {} # Optional, when empty an just a location will be created with no files in it
+  trainer: {} # Optional, will run a training job
+```
+
+### Model Sources
+
+A Model can be sourced from Git or a prebuilt image. Using Git will trigger a container build Job.
+
+```yaml
+kind: Model
+name: falcon-7b
+spec:
+  container:
+    git:
+      url: https://github.com/substratusai/hf-model-loader
+    image: substratusai/hf-model-loader
+  loader:
+    params: {name: tiuu/falcon-7b}
+    # optional future
+    paramsFrom: configMap?
+```
+
+`params` get converted to environment variables following a specific scheme
+
+The controller will orchestrate the following flow in this case:
+
+<img src="./diagrams/model-building.excalidraw.png" width="80%"></img>
+
+
+### Model Training
+
+Approach 1: Include in Model object
+
+Models can be trained by specifying the `.spec.training` section.
+
+This will orchestrate a training Job with multiple FUSE mounted buckets.
+
+<img src="./diagrams/model-training.excalidraw.png" width="80%"></img>
+
+```yaml
+kind: Model
+name: falcon-7b-k8s
+spec:
+  container:
+  # can only specify git or image not both
+    git: 
+      url: https://github.com/substratusai/hf-llm-trainer
+      ref: refs/branch/main
+    image: substratusai/hf-llm-peft-trainer # eventually introduce native pytorch trainer
+
+  trainer:
+    # sourceModel can be left out for creating new base model
+    sourceModel:
+      name: falcon-7b
+      namespace: other-namespace
+    dataset:
+      name: k8s-instructions
+      namespace: other-namespace
+    resources: {}
+    params: {epochs: 1}
+```
+
+
+
+### Model Training with Notebook
+Notebook container contract is to run notebook.sh and serve
+notebook compatible HTTP endpoint on port 8888.
+
+User flow: Random experiments
+
+notebook-gpu image has transformers, pytorch, cuda, python 3, etc.. basically most you need
+
+example notebook in notebook.
+
+1. Create Notebook with no sourceModel
+2. Load a model dynamically in my notebook session
+3. Experiment with the model and training and only save my notebook somewhere for reference
+
+```yaml
+kind: Notebook
+name: notebooks-for-anything
+spec:
+  container:
+    # or git repo as a source
+    image: substratusai/notebook-gpu
+  resources: {gpuCount: 4, type: L4}
+```
+
+
+User flow: Iterate on model training code
+
+1. `git clone https://substratusai/hf-llm-trainer && cd hf-llm-trainer`
+2. Modify Dockerfile
+2. Create a Notebook that mounts source model
+3. `kubectl open notebook notebook-training-experiment`
+   * tar up directory respecting dockerignore
+   * Get GCS signed url from controller notebook status
+   * upload to workspace.tar to signed URL
+3. Notebook automatically has all my local directory in the notebook when using the kubectl plugin
```

Comment:
This probably looks different in the VSCode extension-based flow but using `jupyter lab`:

```suggestion
1. `git clone https://substratusai/hf-llm-trainer && cd hf-llm-trainer`
2. Modify Dockerfile
3. Create a Notebook manifest and apply the Notebook CR mounting a source model
4. `kubectl open notebook -f notebook-training-experiment` - the `kubectl` plugin does each of the following additional steps when `spec.container.local = true`
   * tar up directory respecting dockerignore
   * Get GCS signed url from controller notebook status 
   * upload to workspace.tar to signed URL
5. Notebook automatically has the local directory contents in the notebook when using the `kubectl` plugin (via cp sync operations)
```